File: src/core/time.ts
/** How the timestamp of an entry is written. */
export type TimestampFormat = 'time' | 'datetime' | 'iso' | ((time: number) => string);

const pad = (value: number, length = 2): string => {
	return String(value).padStart(length, '0');
};

/**
 * Formats an epoch time in local time.
 *
 * - `time`: `14:03:09.120`
 * - `datetime`: `2026-09-13 14:03:09.120`
 * - `iso`: `2026-09-13T05:03:09.120Z`, in UTC
 */
export const formatTimestamp = (time: number, format: TimestampFormat = 'time'): string => {
	if (typeof format === 'function') {
		return format(time);
	}

	const date = new Date(time);

	if (format === 'iso') {
		return date.toISOString();
	}

	const clock = `${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}.${pad(date.getMilliseconds(), 3)}`;

	if (format === 'datetime') {
		return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())} ${clock}`;
	}

	return clock;
};

File: packages/flutter/lib/src/core/time.dart
/// How the timestamp of an entry is written.
enum TimestampFormat {
  /// `14:03:09.120`
  time,

  /// `2026-09-13 14:03:09.120`
  datetime,

  /// `2026-09-13T05:03:09.120Z`, in UTC
  iso,
}

String _pad(int value, [int length = 2]) => value.toString().padLeft(length, '0');

/// Formats a time in local time, or in UTC for [TimestampFormat.iso].
String formatTimestamp(DateTime time, [TimestampFormat format = TimestampFormat.time]) {
  if (format == TimestampFormat.iso) {
    return time.toUtc().toIso8601String().replaceFirst(RegExp(r'(\.\d{3})\d*Z?$'), r'$1Z');
  }

  final DateTime local = time.toLocal();
  final String clock =
      '${_pad(local.hour)}:${_pad(local.minute)}:${_pad(local.second)}'
      '.${_pad(local.millisecond, 3)}';

  if (format == TimestampFormat.datetime) {
    return '${local.year}-${_pad(local.month)}-${_pad(local.day)} $clock';
  }

  return clock;
}
